### PR TITLE
Refactor/rule cnvkit paired

### PR DESCRIPTION
**Pull request by ashwini06**
### This PR:
- Splits `CNVkit batch` command into individual commands
- Calculates tumor ploidy and purity using `PURECN`

### Review and tests: 
- [ ] Tests pass - no cnvkit T-N cases for testing 
- [ ] Code review
- [ ] New code is executed and covered by tests, and test approve

### CODE REVIEW

**Review by khurrammaqbool — COMMENTED** [review]
It would be nice to add the screenshots. Is it working/not-crashing given all the possible conditions at config?

**Comment by ashwini06** [comment]
> It would be nice to add the screenshots. Is it working/not-crashing given all the possible conditions at config?

This rule is picked by Jobs and the analysis runs. The problem is that there is no real data for testing TN-cnvkit and with the dummy data, there are cnv variants called and the analysis terminates. 

Eg: for details  check this logs: `/home/proj/production/cancer/github_issue_analysis/issue_40/analysis_APJ/stablehyena_kindkrill/logs.16/BALSAMIC.stablehyena_kindkrill.cnvkit_paired.0.sh_1613233.out` and `/home/proj/production/cancer/github_issue_analysis/issue_40/analysis_APJ/stablehyena_kindkrill/logs.16/BALSAMIC.stablehyena_kindkrill.cnvkit_paired.0.sh_1613233.err`

Screenshot 2021-08-11 at 15 35 05
Screenshot 2021-08-11 at 15 35 35

File: BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule [hunk 1]
```diff
@@ -1,83 +1,163 @@
...
+--output {params.cnv_dir}/normalReference.cnn;
 
-cnvkit.py batch {input.bamT} \
---normal {input.bamN} \
-{params.extra} \
---output-reference {params.cnv_dir}/Reference.cnn \
---scatter --diagram \
---output-dir {params.cnv_dir};
+## Combine the uncorrected target and antitarget coverage tables (.cnn) and
+### correct for biases in regional coverage and GC content, according to the given normal reference
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule:88]:
```suggestion
# Combine the uncorrected target and antitarget coverage tables (.cnn) and
# correct for biases in regional coverage and GC content, according to the given normal reference
```

File: BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule [hunk 2]
```diff
@@ -1,83 +1,163 @@
...
+cnvkit.py fix {params.cnv_dir}/tumor.targetcoverage.cnn \
+{params.cnv_dir}/tumor.antitargetcoverage.cnn \
+{params.cnv_dir}/normalReference.cnn \
+--output {output.cnr};
+       
+## Infer copy number segments from the given coverage table
+### segmentattion methods (-m): cbs: reccommended for mid-size target panels and exomes
+### drop bins with 0  read depth and bins whose log2 values deviating from avg
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule:96]:
```suggestion
# Infer copy number segments from the given coverage table
# segmentattion methods (-m): cbs: reccommended for mid-size target panels and exomes
# drop bins with 0  read depth and bins whose log2 values deviating from avg
```

File: BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule [hunk 3]
```diff
@@ -1,83 +1,163 @@
...
+cnvkit.py segment {output.cnr} \
+--output {params.cnv_dir}/tumor.initial.cns \
+--method cbs \
+--drop-low-coverage \
+--processes {threads};
 
+## Convert copy number segments (.cns) to standard SEG format to be used in PureCN
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule:103]:
```suggestion
# Convert copy number segments (.cns) to standard SEG format to be used in PureCN
```

File: BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule [hunk 4]
```diff
@@ -1,83 +1,163 @@
...
+cnvkit.py export seg {params.cnv_dir}/tumor.initial.cns \
+--output {params.cnv_dir}/tumor.seg;
+
+## Run PureCN to estimate tumor-purity and ploidy
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule:107]:
```suggestion
# Run PureCN to estimate tumor-purity and ploidy
```

File: BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule [hunk 5]
```diff
@@ -1,83 +1,163 @@
...
+--genome hg19 \
+--funsegmentation Hclust \
+--force --postoptimize \
+--seed 124;
+
+purity=$(awk -F\, 'NR>1 {{print $2}}' {params.purecn_dir}/{params.sample_id}.csv)
+ploidy=$(awk -F\, 'NR>1 {{printf int($3)}}' {params.purecn_dir}/{params.sample_id}.csv)
+gender=$(awk -F\, 'NR>1 {{print $4}}' {params.purecn_dir}/{params.sample_id}.csv | tr '[:upper:]' '[:lower:]' | tr -d \\")
+
+## Call copy number variants from segmented log2 ratios
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule:125]:
```suggestion
# Call copy number variants from segmented log2 ratios
```

File: BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule [hunk 6]
```diff
@@ -1,83 +1,163 @@
...
+cnvkit.py call {params.cnv_dir}/tumor.initial.cns \
+--vcf {input.snv_vcf} \
+--sample-sex $gender \
+--method clonal \
+--purity $purity \
+--ploidy $ploidy \
+--output {output.cns};
+
+## Plot bin-level log2 coverages and segmentation calls
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule:134]:
```suggestion
# Plot bin-level log2 coverages and segmentation calls
```

File: BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule [hunk 7]
```diff
@@ -1,83 +1,163 @@
...
+cnvkit.py scatter {output.cnr} \
+--segment {output.cns} \
+--output {output.scatter};
+
+## Draw copy number (.cnr or .cns) on chromosomes as an ideogram
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule:139]:
```suggestion
# Draw copy number (.cnr or .cns) on chromosomes as an ideogram
```

File: BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule [hunk 8]
```diff
@@ -1,83 +1,163 @@
...
+cnvkit.py diagram {output.cnr} \
+--segment  {output.cns} \
+--output {output.diagram};
+
+## Identify targeted genes with copy number gain or loss
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule:144]:
```suggestion
# Identify targeted genes with copy number gain or loss
```

File: BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule [hunk 9]
```diff
@@ -1,83 +1,163 @@
...
 cnvkit.py genemetrics {output.cnr} \
--s {output.cns} \
---drop-low-coverage -y \
+--segment {output.cns} \
+--drop-low-coverage \
+--sample-sex $gender \
 --output {output.gene_metrics};
 
+## List the genenames that contain a possibe copy number breakpoint.
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule:151]:
```suggestion
# List the genenames that contain a possibe copy number breakpoint.
```

File: BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule [hunk 10]
```diff
@@ -1,83 +1,163 @@
...
 cnvkit.py breaks {output.cnr} {output.cns} \
-| cut  -f1 | sort -u > {output.gene_breaks};
+| cut -f1 | sort -u > {output.gene_breaks};
 
+## Convert segments to a vcf file
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule:155]:
```suggestion
# Convert segments to a vcf file
```

**Review by hassanfa — APPROVED** [review]
👍🏽  some format comments, other was A-ok!

File: BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule (hunk 1 above)
Comment (ashwini06) [reply BALSAMIC/snakemake_rules/variant_calling/cnvkit_paired.rule]: Fixed!

**Pull request merged**